engine/crates/composition/src/compose/entity_interface.rs

```diff
@@ -0,0 +1,181 @@
+        if !definition.is_interface_object() {
+            ctx.diagnostics.push_fatal(format!(
+                "`{}` is an entity interface but the object type `{}` is missing the @interfaceObject directive in the `{}` subgraph.",
+                definition.name().as_str(),
+                definition.name().as_str(),
+                definition.subgraph().name().as_str(),
+            ));
+        }
+
+        if definition.entity_keys().next().is_none() {
+            ctx.diagnostics.push_fatal(format!(
+                "The object type `{}` is annotated with @interfaceObject but missing a key in the `{}` subgraph.",
+                interface_name.as_str(),
+                definition.subgraph().name().as_str(),
+            ));
+        }
+
+        for entity_key in definition.entity_keys().filter(|key| key.is_resolvable()) {
+            ctx.insert_interface_resolvable_key(interface_id, entity_key.id, true);
+        }
+
+        for field in definition.fields() {
+            fields.entry(field.name().id).or_insert_with(|| ir::FieldIr {
+                parent_name: definition.name().id,
+                field_name: field.name().id,
+                field_type: field.r#type().id,
+                arguments: field
+                    .arguments()
+                    .map(|arg| (arg.argument_name().id, arg.argument_type().id))
+                    .collect(),
+                resolvable_in: Some(graphql_federated_graph::SubgraphId(definition.subgraph().id.idx())),
+                provides: Vec::new(),
+                requires: Vec::new(),
+                composed_directives: Vec::new(),
+            });
+        }
+    }
+
+    // Contribute the interface fields from the interface object definitions to the implementer of
+    // that interface.
+    for object in iface_def.subgraph().interface_implementers(interface_name.id) {
+        match object.entity_keys().next() {
+            Some(key) if key.fields() == expected_key.fields() => (),
+            Some(_) => ctx.diagnostics.push_fatal(format!(
+                "[{}] The object type `{}` is annotated with @interfaceObject but has a different key than the entity interface `{}`.",
+                object.subgraph().name().as_str(),
+                object.name().as_str(),
+                interface_name.as_str(),
+            )),
+            None => ctx.diagnostics.push_fatal(format!(
+                "[{}] The object type `{}` is annotated with @interfaceObject but missing a key.",
+                object.subgraph().name().as_str(),
+                object.name().as_str(),
+            )),
+        }
+
+        for ir::FieldIr {
+            parent_name: _,
+            field_name,
+            field_type,
+            arguments,
+            resolvable_in,
+            provides,
+            requires,
+            composed_directives,
+        } in fields.values()
+        {
+            ctx.insert_field(ir::FieldIr {
+                parent_name: object.name().id,
```

Comment:
A spread might be cleaner here

Reply:
Will try

Reply:
Spreading is tricky here because of the dereferencing and the `.clone()`s.